File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
            int len=0;
            ListNode tmp=head;

            while(tmp!=null)
            {
                len++;
                tmp=tmp.next;
            }
            int index=len-n-1;
            if(len==n) return head.next;
            tmp=head;
            while(index!=0)
            {   
                index--;
                tmp=tmp.next;
            }
            tmp.next=tmp.next.next;


            return head;
    }
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

def reverse(head):

    curr = head
    prev = None

    while curr:

        nex = curr.next
        curr.next = prev
        prev = curr
        
        curr = nex
    
    return prev

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:

        dummy = ListNode(0)
        dummy.next = head

        i = 0
        curr,temp = head,head
        end = None
        while i < k and curr:
            i+=1
            if i==1:
                start = temp
            if i==k:
                temp = curr.next
                curr.next = None
                if end is None:
                    dummy.next = reverse(start)    
                else:
                    end.next = reverse(start)
                end = start
                end.next = temp
                curr = temp
                i=0
            # print(i , curr.val)
            else:
                curr = curr.next
            
            
        return dummy.next


        